File: frontend/js/producer_consumer.js
let historyData = [];

let currentStep = 0;

let speed = 500;


async function runSimulation(){

    currentStep = 0;

    speed =
        parseInt(
            document.getElementById(
                "speed"
            ).value
        );

    const data = {

        buffer_size:
            parseInt(
                document.getElementById(
                    "bufferSize"
                ).value
            ),

        producer_count:
            parseInt(
                document.getElementById(
                    "producerCount"
                ).value
            ),

        consumer_count:
            parseInt(
                document.getElementById(
                    "consumerCount"
                ).value
            ),

        consume_n:
            parseInt(
                document.getElementById(
                    "consumeN"
                ).value
            ),

        steps:
            parseInt(
                document.getElementById(
                    "steps"
                ).value
            )
    };

    const response = await fetch(

        "http://127.0.0.1:5000/producer_consumer/run",

        {
            method:"POST",

            headers:{
                "Content-Type":
                    "application/json"
            },

            body:
                JSON.stringify(data)
        }
    );

    const result =
        await response.json();

    historyData =
        result.history;

    document.getElementById(
        "log"
    ).innerHTML = "";

    playAnimation();
}


function playAnimation(){

    if(
        currentStep >=
        historyData.length
    ){
        return;
    }

    const item =
        historyData[currentStep];

    renderBuffer(
        item.buffer,
        item.type
    );

    addLog(
        item.action
    );

    currentStep++;

    setTimeout(
        playAnimation,
        speed
    );
}


function renderBuffer(
    buffer,
    type
){

    const area =
        document.getElementById(
            "bufferArea"
        );

    area.innerHTML = "";

    buffer.forEach(value=>{

        const div =
            document.createElement(
                "div"
            );

        div.classList.add(
            "cell"
        );

        if(value===1){

            div.classList.add(
                "full"
            );

            div.innerText="1";

        }else{

            div.classList.add(
                "empty"
            );

            div.innerText="0";
        }

        area.appendChild(
            div
        );
    });
}


function addLog(text){

    const log =
        document.getElementById(
            "log"
        );

    log.innerHTML +=
        text + "<br>";

    log.scrollTop =
        log.scrollHeight;
}

File: backend/algorithms/producer_consumer.py
from flask import Blueprint
from flask import request
from flask import jsonify

producer_consumer_bp = Blueprint(
    "producer_consumer",
    __name__
)


@producer_consumer_bp.route(
    "/producer_consumer/run",
    methods=["POST"]
)
def run_simulation():
    data = request.json
    buffer_size = int(data.get("buffer_size", 10))
    producer_count = int(
        data.get("producer_count", 2)
    )
    consumer_count = int(
        data.get("consumer_count", 2)
    )
    consume_n = int(
        data.get("consume_n", 2)
    )
    steps = int(
        data.get("steps", 20)
    )
    buffer = [0] * buffer_size
    count = 0
    history = []
    producer_index = 0
    consumer_index = 0
    for step in range(steps):
        # 生产阶段
        if step % 2 == 0:
            if count < buffer_size:
                buffer[count] = 1
                count += 1
                producer_id = (
                    producer_index % producer_count
                ) + 1
                producer_index += 1
                history.append({
                    "action":
                        f"Producer {producer_id} Produce",
                    "buffer":
                        buffer.copy(),
                    "type":
                        "produce"
                })
        # 消费阶段
        else:
            if count >= consume_n:
                consumer_id = (
                    consumer_index % consumer_count
                ) + 1
                consumer_index += 1
                for _ in range(consume_n):
                    count -= 1
                    buffer[count] = 0
                history.append({
                    "action":
                        f"Consumer {consumer_id} Consume {consume_n}",
                    "buffer":
                        buffer.copy(),
                    "type":
                        "consume"
                })
    return jsonify({
        "history": history
    })